juju/juju
# AddPendingResources charhub

This PR is a stepping stone to getting resources for charmhub charms in juju.

It implements AddPendingResources for a CharmHub charm.  A charm origin has been added to the args to ensure that the correct channel and architecture are used to find the charm resources.

Additional work is needed to appropriately resolve which resource to use given data from the user, charmhub and the juju db.  As well as upload the resource.

Clean up will be required once the initial work is done.  There are some odd abstractions as the CharmStore client lives outside of juju and the CharmHub client lives inside.  The new charmHubClient inside of the resources facade does some of the work on the charmstore package inside of juju.  The same methods are not required to "resolve" resources between the two repositories: charmstore deals with revisions and charmhub is focused on channels.

## QA steps

Verify that charmstore and local resources have not been broken.

```console
$ export JUJU_DEV_FEATURE_FLAGS="charm-hub"
$ juju bootstrap localhost testme
$ juju add-model six --config charm-hub-url="https://api.staging.snapcraft.io"
$ juju deploy postgresql
Located charm "postgresql" in charm-hub, revision 208
Deploying "postgresql" from charm-hub charm "postgresql", revision 208 in channel latest/stable
```

Check the juju-db to see postgresql resources.  There is no finger print, as download is pending.
